### PR TITLE
[besu] cert-manager setup for besu

### PR DESCRIPTION
**Pull request by RaviKiranJanjanam**
**Changelog**

As of now this is valid only for besu network

added the cert manager helm chart
added new filed in network-besu.yaml "issuer: default/letsencrypt"
updated the code to deploy the cert manager and install certificates using cert-manager using letsencrypt CA, this will run only if "issuer: letsencrypt" in network-besu.yaml, else self signed certs will be created as earlier
updated the reset-network.yaml file to delete the certs at the time of network deletion

 

**Reviewed by**
@sownak @suvajit-sarkar @jagpreetsinghsasan 

 

**Linked issue**
#2036

### CODE REVIEW

File: platforms/hyperledger-besu/configuration/roles/delete/certificates/ambassador/tasks/main.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+- name: delete Ambassador certificates
+  include_tasks: nested_main.yaml
+  vars:
+    node_name: "{{ node.name | lower }}"
+  loop: "{{ services.peers is defined | ternary(services.peers, services.validators) }}"
+  loop_control:
+    loop_var: node
```
Comment (suvajit-sarkar) [inline platforms/hyperledger-besu/configuration/roles/delete/certificates/ambassador/tasks/main.yaml:8]: Fix newline

File: platforms/hyperledger-besu/configuration/reset-network.yaml [hunk 2]
```diff
@@ -26,6 +26,22 @@
         kubernetes: "{{ item.k8s }}"
       loop: "{{ network['organizations'] }}"
 
+    ## Delete ambassador tls certificate for node
+    - name: "delete ambassador tls certificates for Nodes"
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-besu/configuration/reset-network.yaml:30]: Please rename it to `delete ambassador tls certificates for organizations`

File: platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml [hunk 3]
```diff
@@ -16,6 +16,38 @@
...
+      metadata:
+        annotations:
+          cert-manager.io/issue-temporary-certificate: "true"
+          acme.cert-manager.io/http01-edit-in-place: "true"
+        name: "{{ node_name }}-ambassador-certs"
+        namespace: "{{ component_ns }}"
+      spec:
+        secretName: "{{ node_name }}-ambassador-certs"
+        issuerRef:
+          #name: letsencrypt-prod-{{ component_ns }}
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml:36]: Please remove commented code

File: platforms/shared/configuration/roles/setup/certmanager/.travis.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+---
```
Comment (jagpreetsinghsasan) [inline platforms/shared/configuration/roles/setup/certmanager/.travis.yml:1]: Please remove this unwanted file

File: platforms/shared/configuration/roles/setup/certmanager/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
+Role Name
```
Comment (jagpreetsinghsasan) [inline platforms/shared/configuration/roles/setup/certmanager/README.md:1]: Please remove this unwanted file

File: platforms/shared/configuration/roles/setup/certmanager/meta/main.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
+galaxy_info:
```
Comment (jagpreetsinghsasan) [inline platforms/shared/configuration/roles/setup/certmanager/meta/main.yml:1]: Please fix this file to include only the relevant vars and their corresponding comments.

File: platforms/shared/configuration/roles/setup/certmanager/tasks/create_mapping.yaml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+---
```
Comment (jagpreetsinghsasan) [inline platforms/shared/configuration/roles/setup/certmanager/tasks/create_mapping.yaml:1]: The resources created via this file are not idempotent.
Wherever we are supposed to do any change to the current state of the the environment (in this file, the k8s resources), we should first check the current state and perform operations only when the desired state is not same as the current state.
Thus, all the tasks should have a precheck of the existence of the resource and the resource should only be deployed if the pre-check finds the absence of the resource.

File: platforms/shared/configuration/roles/setup/certmanager/tasks/deploy_issuer.yaml [hunk 8]
```diff
@@ -0,0 +1,52 @@
+---
```
Comment (jagpreetsinghsasan) [inline platforms/shared/configuration/roles/setup/certmanager/tasks/deploy_issuer.yaml:1]: The resources created via this file are not idempotent.
Wherever we are supposed to do any change to the current state of the the environment (in this file, the k8s resources), we should first check the current state and perform operations only when the desired state is not same as the current state.
Thus, all the tasks should have a precheck of the existence of the resource and the resource should only be deployed if the pre-check finds the absence of the resource.

**Review by jagpreetsinghsasan — CHANGES_REQUESTED** [review]
1. Certain files in the PR do not have the licence patch, please add them.
2. There are lot of other places where these changes need to be done, and either having those changes in this PR itself or having issues in place for them would keep a better track of the `certmanager` feature. The following 2 follow up issues must be created
- [ ]  Implement cert-manager for besu operational tasks
- [ ] Decoupling of the various resource creation done for cert-manager, from ansible
3. Once the changes are made, please click on `re-request review` so that I will be notified that the PR is now ready for re-review.

File: platforms/shared/configuration/kubernetes-env-setup.yaml [hunk 9]
```diff
@@ -46,6 +46,9 @@
       kubeconfig_path: "{{ item.k8s.config_file }}"
       kubecontext: "{{ item.k8s.context }}"
       aws: "{{ item.aws }}"
+      email: "{{ item.gitops.email }}"
+      organization: "{{ item }}"
+      component_ns: "{{ item.name | lower }}-bes"
```
Comment (suvajit-sarkar) [inline platforms/shared/configuration/kubernetes-env-setup.yaml:51]: the component_ns variable should not be suffixed with bes here

File: platforms/shared/configuration/roles/setup/certmanager/tasks/install_certmanager.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    - molecule-idempotence-notest
+
+##installs cert-manager if its not installed
+- name: Install cert-manager
+  kubernetes.core.helm:
+    kubeconfig: "{{ kubeconfig_path }}"
+    name: cert-manager
+    namespace: cert-manager
+    state: present
+    chart_ref: "{{ playbook_dir }}/../../shared/charts/cert-manager"
```
Comment (suvajit-sarkar) [inline platforms/shared/configuration/roles/setup/certmanager/tasks/install_certmanager.yaml:20]: better to package the cert-manager chart and use it

File: platforms/shared/configuration/roles/setup/certmanager/tasks/create_mapping.yaml [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+##create mapping for acme
+- name: create Mapping for Acme
+  kubernetes.core.k8s:
```
Comment (suvajit-sarkar) [inline platforms/shared/configuration/roles/setup/certmanager/tasks/create_mapping.yaml:4]: see if it makes sense to use move the service mapping and cluster issuer as helm chart

**Review by suvajit-sarkar — CHANGES_REQUESTED** [review]
review comments added; moving it to draft. Kindly move it back to review once ready to be reviewed again

File: platforms/hyperledger-besu/configuration/roles/delete/certificates/ambassador/tasks/main.yaml (hunk 1 above)
Comment (RaviKiranJanjanam) [reply platforms/hyperledger-besu/configuration/roles/delete/certificates/ambassador/tasks/main.yaml]: fixed

File: platforms/hyperledger-besu/configuration/reset-network.yaml (hunk 2 above)
Comment (RaviKiranJanjanam) [reply platforms/hyperledger-besu/configuration/reset-network.yaml]: renamed, licence patch added to all files, cert-manager chart has been packaged

File: platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml (hunk 3 above)
Comment (RaviKiranJanjanam) [reply platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml]: fixed

File: platforms/shared/configuration/roles/setup/certmanager/.travis.yml (hunk 4 above)
Comment (RaviKiranJanjanam) [reply platforms/shared/configuration/roles/setup/certmanager/.travis.yml]: removed

File: platforms/shared/configuration/roles/setup/certmanager/meta/main.yml (hunk 6 above)
Comment (RaviKiranJanjanam) [reply platforms/shared/configuration/roles/setup/certmanager/meta/main.yml]: removed meta/main.yml

File: platforms/shared/configuration/roles/setup/certmanager/tasks/create_mapping.yaml (hunk 7 above)
Comment (RaviKiranJanjanam) [reply platforms/shared/configuration/roles/setup/certmanager/tasks/create_mapping.yaml]: added the code to check for resources prior to installation

File: platforms/shared/configuration/roles/setup/certmanager/tasks/deploy_issuer.yaml (hunk 8 above)
Comment (RaviKiranJanjanam) [reply platforms/shared/configuration/roles/setup/certmanager/tasks/deploy_issuer.yaml]: added the code to check for resources prior to installation

File: platforms/shared/configuration/kubernetes-env-setup.yaml (hunk 9 above)
Comment (RaviKiranJanjanam) [reply platforms/shared/configuration/kubernetes-env-setup.yaml]: corrected

File: platforms/shared/configuration/roles/setup/certmanager/tasks/install_certmanager.yaml (hunk 10 above)
Comment (RaviKiranJanjanam) [reply platforms/shared/configuration/roles/setup/certmanager/tasks/install_certmanager.yaml]: packaged

**Comment by sownak** [comment]
@RaviKiranJanjanam can this be finalised?

File: platforms/hyperledger-besu/configuration/roles/delete/certificates/ambassador/tasks/main.yaml [hunk 12]
```diff
@@ -0,0 +1,23 @@
+##############################################################################################
+#  Copyright Accenture. All Rights Reserved.
+#
+#  SPDX-License-Identifier: Apache-2.0
+##############################################################################################
+---
+- name: delete Ambassador certificates
```
Comment (suvajit-sarkar) [inline platforms/hyperledger-besu/configuration/roles/delete/certificates/ambassador/tasks/main.yaml:7]: role name start with capital to keep it consistent

**Pull request closed without merging**